### PR TITLE
text_processor

### PR DESCRIPTION
**Pull request by Akan186**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_text_processor.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+Description: This module tests the functions in the text_processor module, including counting vowels, reversing text, and converting text to uppercase.
+
+Created on: 04/01/2025
+# spell-checker: disable
+@author: Thilakan Jegatheeswaran
+# spell-checker: enable
+"""
+
+
```
Comment (MD-Jubayer-Khan) [inline solutions/tests/test_text_processor.py:12]: should have a wrapper class here

**Review by MD-Jubayer-Khan — CHANGES_REQUESTED** [review]
"There are no `print` statements anywhere" You can't use print statements. Please remove it

**Comment by Akan186** [comment]
I have made changes. Please review and let me know if anything needs change.

File: solutions/tests/test_text_processor.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    assert results == [
+        "HELLO",
+        "PYTHON",
+        "12345",
+        "!@#$",
+        "AEIOU",
+        " ",
+    ], f"Expected ['HELLO', 'PYTHON', '12345', '!@#$', 'AEIOU', ' '], but got {results}"
+    return results
+
```
Comment (MD-Jubayer-Khan) [inline solutions/tests/test_text_processor.py:119]: Add tests for defensive assertions: Include tests to check the function's behavior with non-string inputs (e.g., integers, lists, None).

File: solutions/text_processor.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    Examples:
+        >>> count_vowels("hello")
+        2
+
+        >>> count_vowels("world")
+        1
+    """
+    vowels = "aeiouAEIOU"
+    return sum(1 for char in text if char in vowels)
+
```
Comment (MD-Jubayer-Khan) [inline solutions/text_processor.py:38]: This function doesn't have any explicit assertions.

File: solutions/text_processor.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    Examples:
+        >>> reverse_text("hello")
+        'olleh'
+
+        >>> reverse_text("world")
+        'dlrow'
+    """
+    return text[::-1]
+
```
Comment (MD-Jubayer-Khan) [inline solutions/text_processor.py:64]: This function doesn't have any explicit assertions.

File: solutions/text_processor.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        str: The text in uppercase.
+
+    Examples:
+        >>> to_uppercase("hello")
+        'HELLO'
+
+        >>> to_uppercase("world")
+        'WORLD'
+    """
+    return text.upper()
```
Comment (MD-Jubayer-Khan) [inline solutions/text_processor.py:89]: This function doesn't have any explicit assertions.

Comment (MD-Jubayer-Khan) [inline solutions/text_processor.py:89]: do something to validate the input
  if not isinstance(text, str):
        raise TypeError("The input must be a string.")

File: solutions/tests/test_text_processor.py (hunk 2 above)
Comment (MD-Jubayer-Khan) [inline solutions/tests/test_text_processor.py:119]: Add tests for boundary cases: Include tests for very long strings, strings with special characters (e.g., punctuation, whitespace), and potentially very short strings (e.g., one-character strings).

**Review by MD-Jubayer-Khan — CHANGES_REQUESTED** [review]

**Review by MD-Jubayer-Khan — CHANGES_REQUESTED** [review]
almost done. do something to validate the input like this in text_processor.py
if not isinstance(text, str):
raise TypeError("The input must be a string.")

**Review by MD-Jubayer-Khan — APPROVED** [review]
Well done mate you did a great job

**Comment by Akan186** [comment]
Thank you!

**Review by Safa-Saber — COMMENTED** [review]
The review is completed. Well done!
